langserver/symbol.go

```diff
@@ -325,14 +325,28 @@ func (h *LangHandler) handleSymbol(ctx context.Context, conn JSONRPC2Conn, req *
 	return results.Results(), nil
 }
 
+type pkgSymResult struct {
+	ready   chan struct{} // closed to broadcast readiness
+	symbols []lsp.SymbolInformation
+}
+
 // collectFromPkg collects all the symbols from the specified package
 // into the results. It uses LangHandler's package symbol cache to
 // speed up repeated calls.
 func (h *LangHandler) collectFromPkg(ctx context.Context, bctx *build.Context, pkg string, rootPath string, results *resultSorter) {
 	h.pkgSymCacheMu.Lock()
-	pkgSyms, _ := h.pkgSymCache[pkg]
+	res, ok := h.pkgSymCache[pkg]
+	if !ok {
+		res = &pkgSymResult{ready: make(chan struct{})}
+		h.pkgSymCache[pkg] = res
+		defer close(res.ready)
+	}
 	h.pkgSymCacheMu.Unlock()
-	if pkgSyms == nil {
+
+	if ok {
```

Comment:
I think this would be much easier to read if you combine this `if` with the `if` above. Also I would suggest to use a second function and do an early-exit on cache hit.